Subject: [PATCH] - added testid
---
Index: packages/ketcher-react/src/script/ui/views/modal/components/Text/TextButton/TextButton.tsx
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/packages/ketcher-react/src/script/ui/views/modal/components/Text/TextButton/TextButton.tsx b/packages/ketcher-react/src/script/ui/views/modal/components/Text/TextButton/TextButton.tsx
--- a/packages/ketcher-react/src/script/ui/views/modal/components/Text/TextButton/TextButton.tsx	(revision e79de8f076f019e2ff116da3eac50861cd1dc815)
+++ b/packages/ketcher-react/src/script/ui/views/modal/components/Text/TextButton/TextButton.tsx	(date 1752148771448)
@@ -44,6 +44,7 @@
     <button
       className={clsx(styles.textButton, { [styles.isActive]: props.active })}
       title={props.button.command.toLowerCase()}
+      data-testid={props.button.name}
       onMouseDown={(event) => {
         toggleStyle(event, props.button.command);
       }}
Index: packages/ketcher-react/src/script/ui/views/modal/components/Text/FontControl/FontControl.tsx
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/packages/ketcher-react/src/script/ui/views/modal/components/Text/FontControl/FontControl.tsx b/packages/ketcher-react/src/script/ui/views/modal/components/Text/FontControl/FontControl.tsx
--- a/packages/ketcher-react/src/script/ui/views/modal/components/Text/FontControl/FontControl.tsx	(revision e79de8f076f019e2ff116da3eac50861cd1dc815)
+++ b/packages/ketcher-react/src/script/ui/views/modal/components/Text/FontControl/FontControl.tsx	(date 1752148981253)
@@ -65,6 +65,7 @@
   return (
     <div ref={wrapperRef}>
       <button
+        data-testid="font-size-button"
         className={classes.fontBtn}
         onMouseDown={(e) => {
           e.preventDefault();
Index: packages/ketcher-react/src/script/ui/views/modal/components/Text/SpecialSymbols/SpecialSymbolsButton.tsx
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/packages/ketcher-react/src/script/ui/views/modal/components/Text/SpecialSymbols/SpecialSymbolsButton.tsx b/packages/ketcher-react/src/script/ui/views/modal/components/Text/SpecialSymbols/SpecialSymbolsButton.tsx
--- a/packages/ketcher-react/src/script/ui/views/modal/components/Text/SpecialSymbols/SpecialSymbolsButton.tsx	(revision e79de8f076f019e2ff116da3eac50861cd1dc815)
+++ b/packages/ketcher-react/src/script/ui/views/modal/components/Text/SpecialSymbols/SpecialSymbolsButton.tsx	(date 1752148940772)
@@ -59,6 +59,7 @@
     <div onBlur={closeSymbolsList}>
       <button
         title="symbols"
+        data-testid="special-symbols-button"
         onMouseDown={(e) => {
           e.preventDefault();
           setShowSpecialSymbols(!showSpecialSymbols);
