Subject: [PATCH] added test ids to texts
---
Index: packages/ketcher-core/src/application/render/restruct/retext.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/packages/ketcher-core/src/application/render/restruct/retext.ts b/packages/ketcher-core/src/application/render/restruct/retext.ts
--- a/packages/ketcher-core/src/application/render/restruct/retext.ts	(revision 237a61b1665ba1249ffb7e8809a9ad61646a43b8)
+++ b/packages/ketcher-core/src/application/render/restruct/retext.ts	(date 1754480225955)
@@ -173,6 +173,11 @@
             fill: '#000000',
             ...styles,
           });
+        path.node.setAttribute('data-testid', 'text');
+        path.node.setAttribute(
+          'data-text-id',
+          restruct.molecule.texts.keyOf(this.item),
+        );
         path.translateAbs(shiftX, shiftY + (styles.shiftY || 0));
         row.push(path);
         shiftX += path.getBBox().width;
